packages/discovery-provider/plugins/pedalboard/apps/solana-relay/src/routes/relay/relay.ts

```diff
@@ -104,54 +118,75 @@ const sendTransactionWithRetries = async ({
   logger: Logger
 }) => {
   const serializedTx = transaction.serialize()
-  const connection = connections[0]
 
-  const createRetryPromise = async (): Promise<void> => {
-    let retryCount = 0
-    while (true) {
-      await delay(RETRY_DELAY_MS)
-      // Explicitly not awaited, sent in the background
-      logger.info({ retryCount }, 'Attempting send...')
-      try {
-        Promise.any(
-          connections.map((connection) =>
-            connection.sendRawTransaction(serializedTx, sendOptions)
-          )
-        )
-      } catch (error) {
-        logger.warn(
-          { error, retryCount, rpcEndpoint: connection.rpcEndpoint },
-          `Failed retry...`
+  let retryCount = 0
+  const createRetryPromise = async (signal: AbortSignal): Promise<void> => {
+    while (!signal.aborted) {
+      Promise.any(
+        connections.map((connection) =>
+          connection.sendRawTransaction(serializedTx, {
+            skipPreflight: true,
+            maxRetries: 0,
+            ...sendOptions
+          })
         )
-      }
+      ).catch((error) => {
+        logger.warn({ error, retryCount }, `Failed retry...`)
+      })
+      await delay(RETRY_DELAY_MS)
+      retryCount++
     }
   }
 
-  const createTimeoutPromise = async () => {
+  const createTimeoutPromise = async (signal: AbortSignal) => {
     await delay(RETRY_TIMEOUT_MS)
-    logger.error('Timed out sending transaction')
+    if (!signal.aborted) {
+      logger.error('Timed out sending transaction')
+    }
   }
 
   const start = Date.now()
+  const connection = connections[0]
+  const abortController = new AbortController()
+  try {
+    if (!sendOptions?.skipPreflight) {
+      const simulatedRes = await connection.simulateTransaction(transaction)
+      if (simulatedRes.value.err) {
+        logger.error(
+          { error: simulatedRes.value.err },
+          'Transaction simulation failed'
+        )
+        throw simulatedRes.value.err
+      }
+    }
 
-  const res = await Promise.race([
-    createRetryPromise(),
-    connection.confirmTransaction(confirmationStrategy, commitment),
-    createTimeoutPromise()
-  ])
-
-  const end = Date.now()
-  const elapsedMs = end - start
+    const res = await Promise.race([
+      createRetryPromise(abortController.signal),
+      connection.confirmTransaction(
+        { ...confirmationStrategy, abortSignal: abortController.signal },
+        commitment
+      ),
+      createTimeoutPromise(abortController.signal)
+    ])
 
-  if (!res || res.value.err) {
+    if (!res || res.value.err) {
+      throw res?.value.err ?? 'Transaction polling timed out.'
```

Comment:
This feels like it might create false positives about timeouts. Should we not have the timeout promise explicitly return a value here and reserve the null-coalescing for something like "Unknown..." so that we know it's an unhandled case?